### PR TITLE
pointcut exceptions

### PR DESCRIPTION
**Pull request by fil512**
Enforce no Pointcut exceptions are thrown on the invoking thread

### CODE REVIEW

File: hapi-fhir-jpaserver-test-r5/src/test/java/ca/uhn/fhir/jpa/subscription/resthook/RestHookTestR5IT.java [hunk 1]
```diff
@@ -129,7 +134,7 @@ public void testUpdatesHaveCorrectMetadata() throws Exception {
 
 		sentObservation.getIdentifierFirstRep().setSystem("foo").setValue("2");
 		updateResource(sentObservation, true);
-		sentObservation = myObservationDao.read(sentObservation.getIdElement().toUnqualifiedVersionless(), mySrd);
+		sentObservation = myObservationDao.read(sentObservation.getIdElement().toUnqualifiedVersionless());
```
Comment (StevenXLi) [inline hapi-fhir-jpaserver-test-r5/src/test/java/ca/uhn/fhir/jpa/subscription/resthook/RestHookTestR5IT.java:137]: Question: Is there a reason you're using the depreciated version of the read in this test class?

File: hapi-fhir-jpaserver-test-r5/src/test/java/ca/uhn/fhir/jpa/dao/r5/BaseJpaR5Test.java [hunk 2]
```diff
@@ -437,15 +437,15 @@ public void afterClearTerminologyCaches() {
...
 	}
 
 	@BeforeEach
 	public void beforeCreateInterceptor() {
 		myInterceptor = mock(IServerInterceptor.class);
 
-		myPerformanceTracingLoggingInterceptor = new PerformanceTracingLoggingInterceptor();
-		myInterceptorRegistry.registerInterceptor(myPerformanceTracingLoggingInterceptor);
+//		myPerformanceTracingLoggingInterceptor = new PerformanceTracingLoggingInterceptor();
+//		myInterceptorRegistry.registerInterceptor(myPerformanceTracingLoggingInterceptor);
```
Comment (StevenXLi) [inline hapi-fhir-jpaserver-test-r5/src/test/java/ca/uhn/fhir/jpa/dao/r5/BaseJpaR5Test.java:448]: Nitpick: should these commented out code just be removed? are we going to need the PerformanceTracingLoggingInterceptor again in the future here?

File: hapi-fhir-jpaserver-test-r5/src/test/java/ca/uhn/fhir/jpa/subscription/resthook/RestHookTestR5IT.java (hunk 1 above)
Comment (fil512) [reply hapi-fhir-jpaserver-test-r5/src/test/java/ca/uhn/fhir/jpa/subscription/resthook/RestHookTestR5IT.java:137]: I can't find that change.  Did you review a different merge?

File: hapi-fhir-jpaserver-test-r5/src/test/java/ca/uhn/fhir/jpa/dao/r5/BaseJpaR5Test.java (hunk 2 above)
Comment (fil512) [reply hapi-fhir-jpaserver-test-r5/src/test/java/ca/uhn/fhir/jpa/dao/r5/BaseJpaR5Test.java:448]: Godo point, but I think it's helpful to leave junk like this in test classes so we can easily switch them on and off

File: hapi-fhir-jpaserver-test-r5/src/test/java/ca/uhn/fhir/jpa/subscription/resthook/RestHookTestR5IT.java (hunk 1 above)
Comment (fil512) [reply hapi-fhir-jpaserver-test-r5/src/test/java/ca/uhn/fhir/jpa/subscription/resthook/RestHookTestR5IT.java:137]: i see it now.  fixed.

File: hapi-fhir-test-utilities/src/main/java/ca/uhn/test/concurrency/PointcutLatchSession.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+public class PointcutLatchSession {
```
Comment (michaelabuckley) [inline hapi-fhir-test-utilities/src/main/java/ca/uhn/test/concurrency/PointcutLatchSession.java:16]: one line of javadoc would be kind.

File: hapi-fhir-test-utilities/src/test/java/ca/uhn/test/concurrency/PointcutLatchTest.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.startsWith;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
+import static org.junit.jupiter.api.Assertions.fail;
+
+class PointcutLatchTest {
```
Comment (michaelabuckley) [inline hapi-fhir-test-utilities/src/test/java/ca/uhn/test/concurrency/PointcutLatchTest.java:19]: Nice

**Review by michaelabuckley — APPROVED** [review]

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/interceptor/api/HookParams.java [hunk 5]
```diff
@@ -136,8 +134,14 @@ public <T> HookParams addIfMatchesType(Class<T> theType, Object theParam) {
...
-		return new ToStringBuilder(this, ToStringStyle.SIMPLE_STYLE)
-			.append("params", myParams)
-			.toString();
+		StringBuilder b = new StringBuilder();
+		myParams.forEach((key, value) -> {
+			b.append("  ")
+				.append(key.getSimpleName())
+				.append(": ")
+				.append(value)
+				.append("\n");
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-base/src/main/java/ca/uhn/fhir/interceptor/api/HookParams.java:143]: This is only used in tests.  I needed to read tonnes of these to fix all the latches.

File: hapi-fhir-jpaserver-mdm/src/test/java/ca/uhn/fhir/jpa/mdm/interceptor/MdmSearchExpandingInterceptorIT.java [hunk 6]
```diff
@@ -257,12 +257,11 @@ theDetails, new PatientEverythingParameters(), new IdDt(id)
 	}
 
 	@Test
-	public void testReferenceExpansionQuietlyFailsOnMissingMdmMatches() {
+	public void testReferenceExpansionQuietlyFailsOnMissingMdmMatches() throws InterruptedException {
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-jpaserver-mdm/src/test/java/ca/uhn/fhir/jpa/mdm/interceptor/MdmSearchExpandingInterceptorIT.java:260]: ya I posted about this on slack.  gunna do that cleanup in a separate MR

File: hapi-fhir-test-utilities/src/main/java/ca/uhn/test/concurrency/PointcutLatchSession.java (hunk 3 above)
Comment (fil512) [reply hapi-fhir-test-utilities/src/main/java/ca/uhn/test/concurrency/PointcutLatchSession.java:16]: done

**Pull request merged**